src/compiler/checker.ts

```diff
@@ -28031,7 +28035,22 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             return getNarrowedType(type, instanceType, assumeTrue, /*checkDerived*/ true);
         }
 
-        function getInstanceType(constructorType: Type) {
+        function getInstanceType(constructorType: Type, left: Expression, leftType: Type, right: Expression) {
+            // if rightType is an object type with a custom `[Symbol.hasInstance]` method, and that method has a type
+            // predicate, use the type predicate to perform narrowing. This allows normal `object` types to participate
+            // in `instanceof`, as per Step 2 of https://tc39.es/ecma262/#sec-instanceofoperator.
+            const hasInstanceMethodType = getSymbolHasInstanceMethodOfObjectType(constructorType);
+            if (hasInstanceMethodType) {
+                const syntheticCall = createSyntheticHasInstanceMethodCall(left, leftType, right, constructorType, hasInstanceMethodType);

```

Comment:
I'm surprised at all the complexity with synthetic method calls and so on. Isn't it just a matter of finding the `hasInstance` method, checking that it has a type predicate return type, and fishing the target type out of that. Why do we need synthetic calls and overload resolution?

Reply:
It would be perfectly reasonable to assume that a `[Symbol.hasInstance]()` method could have overloads, just as any function or method with a type predicate. The synthetic call gives us the ability to leverage normal call resolution to be consistent with how the transformed representation would be handled:

```js
x instanceof y
// is equiv. to
y[Symbol.hasInstance](x)
```